idea/src/org/jetbrains/kotlin/idea/editor/KotlinLiteralCopyPasteProcessor.kt

```diff
@@ -0,0 +1,285 @@
+                    }
+
+                    is EntryChunk -> chunk.text
+                    is NewLineChunk -> "\n"
+
+                }
+            }.joinToString(separator = "")
+        }
+
+    }
+
+
+}
+
+private sealed class TemplateChunk
+private data class LiteralChunk(val text: String) : TemplateChunk()
+private data class EntryChunk(val text: String) : TemplateChunk()
+private class NewLineChunk : TemplateChunk()
+
+private class TemplateTokenizer(private val inputString: String) : Sequence<TemplateChunk> {
+
+    private fun String.guessIsTemplateEntryStart(): Boolean =
+            if (this.startsWith("\${")) {
+                true
+            }
+            else if (this.length > 1 && this[0] == '$'){
+                KotlinLexer().apply { start(this@guessIsTemplateEntryStart.substring(1)) }.tokenType == KtTokens.IDENTIFIER
+            }
+            else {
+                false
+            }
+
+    private fun findTemplateEntryEnd(input: String, from: Int): Int{
+        val wrapped = '"' + input.substring(from) + '"'
+        val lexer = KotlinLexer().apply { start(wrapped) }.apply { advance() }
+
+        if (lexer.tokenType == KtTokens.SHORT_TEMPLATE_ENTRY_START){
+            lexer.advance()
+            return if (lexer.tokenType == KtTokens.IDENTIFIER){
+                from + lexer.tokenEnd -1
+            } else{
+                -1
+            }
+        }
+        else if(lexer.tokenType == KtTokens.LONG_TEMPLATE_ENTRY_START) {
+            var depth = 0
+            while (lexer.tokenType != null) {
+                if (lexer.tokenType == KtTokens.LONG_TEMPLATE_ENTRY_START) {
+                    depth++
+                }
+                else if (lexer.tokenType == KtTokens.LONG_TEMPLATE_ENTRY_END) {
+                    depth--
+                    if (depth == 0) {
+                        return from + lexer.currentPosition.offset - 1
+                    }
+                }
+                lexer.advance()
+            }
+            return -1
+        }
+        else {
+            return -1
+        }
+    }
+
+    private suspend fun SequenceBuilder<TemplateChunk>.yieldLiteral(chunk: String) {
+
+            val splitLines = LineTokenizer.tokenize(chunk, false, true)
+            for (i in 0..splitLines.size - 1) {
+                if (i != 0) {
+                    yield(NewLineChunk())
+                }
+                splitLines[i].takeIf { !it.isEmpty() }?.let { yield(LiteralChunk(it)) }
+            }
+
+    }
+
+    private fun iterTemplateChunks(): Iterator<TemplateChunk> {
+        if (inputString.isEmpty()) {
+            return emptySequence<TemplateChunk>().iterator()
+        }
+        return buildIterator {
+            var from = 0
+            var to = 0
+            while (to < inputString.length) {
+                val c = inputString[to]
+                if (c == '\\') {
+                    to += 1
+                    if (to < inputString.length)
+                        to += 1
```

Comment:
Actually, after giving it some though I don't think it will be a problem because in this case we only care about escape sequences which apply to escaping '$'